### DISCUSSION TITLE
Security audit of Reticulum+LXMF

### DISCUSSION THREAD

**Discussion by cannon-c** [markqvist/Reticulum#628] [General]
I'm organizing a security audit of Reticulum+LXMF. 
 
I am in contact with a company called Trail of Bits whom has expertise in conducting security audits of software, protocols, cryptography. Before proceeding, it would be good to understand what the community feels would have the biggest impact from a third-party security review.
 
 1. If we had a third party review by an expert group such as Trail of Bits, what would be the most valuable thing they could help with?
  
 2. Where would people in the community feel like Trail of Bits could help enhance security or make the development processes more secure and efficient?
 
 3. Is it automated testing tools, formal verification, or something else that the community would find most useful?

Any other thoughts or suggestions?

My personal thoughts initially were to request:
* Audit of the cryptography implementation of Reticulum and LXMF
* Software security review of Reticulum, LXMF, NomadNet (emphasis on security from remote code exploits)
   
After having this extra input and answers to these questions from others, Trail of Bits would be happy to setup a call with the broader community to discuss the security review considerations and get feedback.

**Comment by cannon-c** [comment]
Let's use this page for discussing this topic. Please feel free to link to this and make mention of this in other places such as Reddit, Matrix, wherever else appropriate and to forward and consolidate comments from such into here.

**Comment by LoongUser77** [comment]
Hi,
Maybe this helps, I have run Bandit ([https://github.com/PyCQA/bandit]) over nomadnet code:
[Bandit_in_Nomadnet.txt](https://github.com/user-attachments/files/18080925/Bandit_in_Nomadnet.txt)

It won't test crypto security, but just the code. But it could help

**Reply by cannon-c** [reply]
I was not aware of this Bandit tool, this is neat. While not a replacement of course for a manual and professional audit; this is still helpful for finding at least some potential issues. Thank you for sharing these results! This is helpful.

**Comment by silveroranges** [comment]
The things I would love to see audited is essentially your first thought, Reticulum, LXMF and how it implements the cryptography. 

I don't know if this is in the scope of what they do, but also potential flaws that could lead to a DoS attack would also be good to know, in case a well funded group or state tries to block the network. I think this would be a good avenue of investigation because while jamming may work on one specific frequency or even a range of frequencies, the nature of reticulum and how it is able to communicate over any medium would mean the attack of choice would be something against the network specifically and not the transport medium.

**Reply by cannon-c** [reply]
You have a good point, as you do point out, a Reticulum network can overlap across many different mediums making a DoS attack at the software/protocol layer more widespread as a result. Especially since Reticulum is desirable as a solution for sovereign decentralized networks, availability is of high importance.

**Reply by bornac1** [reply]
You can always dinamically change the frequency by shifting it slightly to prevent jamming, but preventing DoS and DDoS is very hard. I would go in a direction of rate-limiting, but maybe something is even better...
Do you have any metrics for processing large number Reticulum packets? Did anybody test it on low-power hardware you are planning to use?

**Comment by LinuxinaBit** [comment]
You’re gonna want to get @markqvist ’s and @faragher ’s opinion on this, but I think you’re absolutely right about auditing the whole cryptography/identity system at _least_.
Probably gonna be hard to get them to audit the protocols themselves because I don’t think that’s what they specialize in, and I don’t think an audit of NomadNet specifically is that important (it’s just one of many LXMF clients), but I do think a general audit of the reference Reticulum and LXMF implementations is a good idea.

**Reply by LinuxinaBit** [reply]
The other thing to worry about is how much this would cost. I’m not terribly familiar with Trail of Bits and their pricing, but I don’t think it’s terribly wise to audit things other than the reference stack (Reticulum and LXMF) for now.

**Reply by bornac1** [reply]
> The other thing to worry about is how much this would cost. I’m not terribly familiar with Trail of Bits and their pricing, but I don’t think it’s terribly wise to audit things other than the reference stack (Reticulum and LXMF) for now.

I don't know for Python, but for C# it starts from 5€ per one line of code for the agency I use. I managed to get lower price (can't say exact publicly) for my project because of MIT licence and my work (my company uses their service as well).

**Comment by markqvist** [comment]
Yes, as @LinuxinaBit says, the essential parts to get verified right now are the core workings of Reticulum itself, in order of priority:

- The core cryptography implementation (not the pure-python stuff for now, we still consider that highly experimental, and is only really there for the sake of experimentation and development)
- The core `Identity` implementation and `Destination` derivation from identity keys and destination aspects
- The cryptography involving the announce mechanism, and signature validation of received announces, plus mapping to valid destination paths
- Link establishment, link key exchange and encryption setup for links
- Ratchet distribution for ratchet-enabled destinations and per-packet ratchet-based encryption

In my personal opinion, those are the most relevant and valuable parts to get verified at this point. Once that is done, it would be interesting to start looking at higher-level protocols such as LXMF, but not before.

**Reply by cannon-c** [reply]
"(not the pure-python stuff for now, we still consider that highly experimental, and is only really there for the sake of experimentation and development)"

@markqvist Is the python cryptography a required dependency of reticulum or LXMF, or is it only optional or if other default dependencies are not present?

**Reply by cannon-c** [reply]
"In my personal opinion, those are the most relevant and valuable parts to get verified at this point. Once that is done, it would be interesting to start looking at higher-level protocols such as LXMF, but not before."

Would a cryptography audit of Reticulum also give assurance of proper cryptography implementation in LXMF or no? I personally consider LXMF of equal importance as Reticulum as for my purposes at least, I am using Reticulum for communications and I personally see LXMF as a core feature of Reticulum.

**Reply by LinuxinaBit** [reply]
For how exactly the Python cryptography implementation is incorporated into Reticulum, it might be worth checking out this page: https://reticulum.network/crypto.html

>Reticulum also includes a complete implementation of all necessary primitives in pure Python. If OpenSSL & PyCA are not available on the system when Reticulum is started, Reticulum will instead use the internal pure-python primitives. A trivial consequence of this is performance, with the OpenSSL backend being much faster. The most important consequence however, is the potential loss of security by using primitives that has not seen the same amount of scrutiny, testing and review as those from OpenSSL.
>
>If you want to use the internal pure-python primitives, it is highly advisable that you have a good understanding of the risks that this pose, and make an informed decision on whether those risks are acceptable to you.

**Comment by bornac1** [comment]
@markqvist Finally you are discussing the idea that I suggested many months ago...

**Reply by markqvist** [reply]
I'm quite unsure what the point of your comment is here, and as such really have no way of answering it.

**Reply by bornac1** [reply]
> I'm quite unsure what the point of your comment is here, and as such really have no way of answering it.

I mentioned this a long time ago while I was sending NoIPChat for certification and I didn't get any answer whether you would like me to send it as well or not. Also, Reticulum was updated too fast at that time and any certificate we would get would be for the outdated version.

**Reply by markqvist** [reply]
> Reticulum was updated too fast at that time and any certificate we would get would be for the outdated version

Exactly.

**Comment by KenSamson** [comment]
I set up an lmxd instance...  set it to announce..   then when it did announce I could see it in both MeshChat and NomadNet as clients, using a shared rnsd instance runng on the same machine.

I then sent it messages... looking to play around with storing messages on the server and writing a program to come retrieve them later.   All experimental stuff.   Just trying to learn.

in the .lxmd/storage/messages folder, a message got stored for each one I sent, in a separate file.    All seems to be working just fine.

Then I opened one of the messages in notepad on windows... and the body of the message was stored in cleartext...

![image](https://github.com/user-attachments/assets/60ee6f8f-a34f-49b3-8bcc-016c947e7b9e)

I was really hoping that the body of the message would have been encrypted in an lmxf message.. but maybe I have a lot more to learn.

It may be that because the server itself is the destination of these test messages (which really is a bit of an unusual circumstance), the server just happens to have the required keys handy and decrypts the message.     I can think of several ways that could be useful where you could send the server commands because it would decode the messages.   Would it not be useful to be able to use the network to send the routers instructions?  (powerful and dangerous too, perhaps.)

To test end to end encryption, I would need to send the lxmd a message that is not "TO" the server itself.    I'll work on that next, and see what the message looks like.

**Reply by markqvist** [reply]
I think you should post this in a new thread, since it is not relevant to the current subject of security auditing of Reticulum, but it is still relevant to an overall understanding of LXMF.

That being said, what you're seeing here is the correct and intended behaviour. As you say, you are sending LXMF messages to the `lxmf.delivery` endpoint of the LXMF instance you are running. The message gets decrypted and handled by whatever functionality you configure. If the final receiver of an LXMF message could not decrypt it and act on it, there would be no point of the protocol. The `lxmd` program can act both as a final receiver of LXMF messages, and as a propagation node for other instances.

If you enable the propagation node functionality of this instance in the LXMF config file, and send a message to *another* destination via this propagation node, it will of course be stored encrypted in the propagation message store. You can verify this for yourself.

**Comment by RFnexus** [comment]
Hi @cannon-c,

I visited the node you're hosting from the address in your Github profile and saw this:
![image](https://github.com/user-attachments/assets/adbdced1-3c9c-48e4-aa98-276e592a8cd2)

Could I ask what this entails, and what the logistics are you have planned for fundraising? I think it is very important to be as transparent as possible with any sort of fundraising effort as this requires a large degree of trust between the organizer and the community. There are far too many bad actors out there when it comes too crowdfunding (see https://blog.thecrowdfundingformula.com/crowdfunding-scams/ for examples) and unfortunately I just don't think this approach is something that can be done realistically.

**Reply by cannon-c** [reply]
I agree that trust should be minimized. And I absolutely agree with the importance of transparency. This is a good place to start discussion on this as well.

"I just don't think this approach is something that can be done realistically." Do you mean the approach of crowdfunding as being unrealistic or something else?

What I have in mind was having funds raised from two ways. Applications submitted to various organizations that give funding and grants to open source projects, and also with a way for anyone to donate to go towards it also. Perhaps funding from such organizations  can be released directly from the one or more pledgers to the company conducting the audit, after the total amount of pledges is sufficient to fund a the security audit. As for donations from people, perhaps a bitcoin wallet under the control of someone who has reputation and merit in this space (not me as I am new and not well known), a multisig would be better to eliminate a sincle point of failure and eliminate the need to trust a single person. Geyser Fund (https://geyser.fund) platform can be looked into (I am personally  starting more research on Geyser Fund), or it can be as simple as a listed BTC address. While lightning transactions are great for micro-donations, onchain may be better suited for this since onchain addresses would support multisig and transparency of all donations. Something to think about, is what to do with the already donated funds in the BTC address if the targetted funding amount is not reached within the targetted timeframe. Would the funds just remain in that donation address allocated for the specific purpose (security audit) until one day when it is a sufficient amount to fund a security audit? Would the funds get refunded (minus the transaction fee) back to the sender after a certain time? Maybe instead of collecting donations from people, have people pledge what they'd be willing to donate and then start collecting donations after the pledged amount is sufficient? But if go that route with _people_ then also need to think about how to prevent abuse of intentionally fake pledges.

Alternatively, if anyone knows how a proper security audit can be conducted without the need of funds (for free), please do share.

**Reply by cannon-c** [reply]
My personal suggestion would be grants from organizations paid directly to the company conducting such an audit conditionally if the audit is collectively fully funded within a declared timeframe; alongside a bitcoin multisig donation address collaboratively controlled by people with merit and reputation in this space; with donations refunded to senders (minus transaction fees) if funding is not reached in the  same timeframe. Let's hear other thoughts.

**Comment by cannon-c** [comment]
After the community confirms the scope of a security audit, we can then confirm with Trail of Bits if everything would be within their scope and capabilities and then can see if Trail of Bits would be able to have a public meeting. Then after a scope is established by both the community and Trail of Bits; then the funding efforts can take place. As of now, this is what I have noted as a proposed scope based on input so far in this discussion, including what markqvist has commented. I have two questions in this post, marked by #?

Start with the scope of the foundational protocol of Reticulum and LXMF.
#? I had some questions I asked in here to markqvist regarding the scope of LXMF in the audit. If there is debate on whether to include LXMF in the scope of the audit or not, maybe Trial of Bits can give a quote on an audit with Reticulum and another with Reticulum+LXMF.

In order of priority:

* The core cryptography implementation ("not the pure-python stuff for now, we still consider that highly experimental, and is only really there for the sake of experimentation and development" markqvist)

#? Would these following points be sub-categories under "core cryptography implementation", or additional categories?

	* The core Identity implementation and Destination derivation from identity keys and destination aspects

	* The cryptography involving the announce mechanism, and signature validation of received announces, plus mapping to valid destination paths

	* Link establishment, link key exchange and encryption setup for links

	* Ratchet distribution for ratchet-enabled destinations and per-packet ratchet-based encryption

* Security from DoS attacks at the protocol level (due to it overlapping with various networks and mediums)

**Reply by LinuxinaBit** [reply]
Auditing security from DoS attacks is not something Trail of Bits is likely going to be able to do (depending on what *exactly* you mean, it could also be physically impossible), and doing so is not of very high importance in Reticulum's case anyways.
See https://github.com/markqvist/Reticulum/wiki/Frequently-Asked-Questions#whats-to-stop-someone-from-flooding-the-network and https://github.com/markqvist/Reticulum/wiki/Frequently-Asked-Questions#whats-to-stop-someone-from-flooding-the-radio-to-prevent-communication for more details.

**Reply by LinuxinaBit** [reply]
I suggest studying up on Reticulum a little bit more before you jump in and start trying to get things audited.
I have a ton of resources linked at the bottom of [my Reticulum explainer](https://linuxinabit.codeberg.page/blog/reticulum) that should help you understand why certain things need to be audited, what they are, and what they do.
Hope that helps a little :)

**Reply by cannon-c** [reply]
@LinuxinaBit 

> I suggest studying up on Reticulum a little bit more before you jump in and start trying to get things audited. I have a ton of resources linked at the bottom of [my Reticulum explainer](https://linuxinabit.codeberg.page/blog/reticulum) that should help you understand why certain things need to be audited, what they are, and what they do. Hope that helps a little :)

I have read alot on Reticulum. Though as you say, I have more to read up on which I am now. Thank you for the information, I am reading up on Reticulum further with your provided references. As for an audit, I have the opinion that such needs done sooner rather than later. If not now, then when? Assurance of security of the protocol would really bolster confidence in it and allow for serious adoption of it for various purposes. And this audit process should be a collective effort to fill knowledge gaps and to not have a single point of failure and trust. And plenty of people here have alot of knowledge on Reticulum. Do you think an audit should be delayed, or do you think the collective effort and knowledge for such is sufficient at this time? Thanks again for your input and suggestions.

**Reply by cannon-c** [reply]
also, nice looking website and presentation of information @LinuxinaBit

**Reply by bornac1** [reply]
Cryptographic primitives are usually audited just once and most likely their original author (company) will send you certificates they have for a small fee or sometimes free.
Or maybe even your auditor has previously used some of these libraries and might already have them. That was the case for NoIPChat as ZSIS already audited many .NET applications.

Hence, I would start from Reticulum protocol itself as a base and assume that libraries are safe. These libraries are huge and it can cost thousands of euros to audit them. And later if you have enough money you can audit libraries as well, replace them with an already audited ones.

**Reply by LinuxinaBit** [reply]
@bornac1 Yes, I believe that is referring to auditing how the cryptography is handled on the Reticulum end of things, not the libraries themselves (which are provided by OpenSSL).

**Reply by LinuxinaBit** [reply]
@cannon-c 
> As for an audit, I have the opinion that such needs done sooner rather than later. If not now, then when? Assurance of security of the protocol would really bolster confidence in it and allow for serious adoption of it for various purposes. And this audit process should be a collective effort to fill knowledge gaps and to not have a single point of failure and trust. And plenty of people here have alot of knowledge on Reticulum. Do you think an audit should be delayed, or do you think the collective effort and knowledge for such is sufficient at this time? Thanks again for your input and suggestions.

I think an audit is absolutely necessary, but not *immediately*. That being said, I do think the process should be started sooner rather than later, because, as I'm sure you understand, it's will take a while to both gain funding, and just go through the process.
Also keep in mind that there seem to be plans to do pretty sizable reworks to the core of Reticulum in the near(ish) future to allow for things like Group Destinations (which have to be route-able and cryptographically sound).

**Reply by cannon-c** [reply]
I was traveling internationally with limited access to internet and just recently got back, so pardon my absence on here. 
I am finally well rested and recovered from my trip.

@LinuxinaBit 

> there seem to be plans to do pretty sizable reworks to the core of Reticulum in the near(ish) future

1. Would such changes render audits already conducted no longer relevant? If so then it may be best to wait till the core protocol is established before proceeding with an audit.

2. Would anyone have an idea of how far in the future it would be for such changes to be applied?

Even if we do not proceed with an audit at this time, I see benefit in formulating a plan now for such an audit that is ready to go for when the time is right. A step by step plan such as scope, who to contact, how to raise funds like which organizations to submit applications to for grants/funding, how to setup a trust-less (multisig wallet?) crowd funding donation address. So that when the time is right, it can be easily and efficiently acted upon. Such a plan would be optional, and can be adopted for the purpose in partial or full as seen fit when appropriate.

**Reply by LinuxinaBit** [reply]
@cannon-c 
All good.
As for your questions:
1. Yes, most likely. Audits generally only apply for a single version of the audited software.
2. No ETAs because they’re never accurate, but the minimum will be to wait for RNS 1.0.0

Mostly it’s just waiting now; funding etc. will be figured out once an audit is actually being worked on.

**Reply by bornac1** [reply]
@cannon-c @LinuxinaBit 
In all cases, certificated obtained after audit are valid only as long as you don't make any change in your software. Starting from the source code up to compiler. Most compilers have proofs that they provide equivalent outputs, so they can be changed. But in all cases, once you change source code even the smallest change, your certificate is legally invalid.

What's the problem with 1.0.0? Are we waiting for some new features?

**Reply by LinuxinaBit** [reply]
> What's the problem with 1.0.0?

I believe you misread.
I said that we should wait until RNS _at least_ hits version 1.0.0

> Are we waiting for some new features?

Yes. Please read my previous messages on this subject.

**Comment by LinuxinaBit** [comment]
I believe it is wholly unproductive to discuss audit details/specifics before Reticulum has even had its first full release.

It is perfectly fine to plan for it, but there are **major reworks** coming in the *near future* that will almost completely nullify the benefits of any premature audit.

It is simply not worth the cost until those reworks actually materialize, and it is far better to fund actual development up until that point.

As for right now, I believe that everything that can currently be discussed has been discussed, and as such ~I will be unsubscribing from this discussion.~
Thank you.

**Reply by faragher** [reply]
I would go so far as to say it's unproductive to hold an audit before an LTS version. Unless there's a reference version which is frozen and intended to last for may years or even decades, there's no point in doing an audit on that version. It needs to be something that is feature complete and can do whatever the end user asks of it, because any updates would then not be certified. This isn't version 1.0.0, this is deep in the bug-fix lifecycle and something considered fully mature. Then it can be frozen and subjected to the most intensive of investigations.

**Comment by cannon-c** [comment]
So it is clear from the input here, that we should wait until Reticulum is finalized in it's base protocol and first full release.

(after the time is right), I currently disagree though that audits are useless anytime the software updates for two reasons:
1. Potential issues effecting the foundational code would be identified and addressed early on.
2. Updates after the point of an audit are easy to vet and monitor since the majority of the code was already audited. If an update is only 5 lines of code, it is not necessary to retroactively audit the tens of thousands of lines of code already audited before that point?

**Reply by cannon-c** [reply]
I can close this discussion out a week from now to allow for other comments (unless someone wishes to close it out before I do). And such discussion can be re-addressed at a later point when the time is right. The conversation in here has still been productive I think, and things brought up in here I feel may be of benefit for such efforts which may arise in the future.

**Reply by LinuxinaBit** [reply]
There are two reasons to audit a piece of software:
1. To find vulnerabilities and issues that need to be fixed.
2. To *certify* that the developers made a secure program designed with, to the best of the auditor's knowledge, no backdoors or vulnerabilities.

When a new version of a piece of software is released, the benefits of point 1 don't go away, but the *certification* of point 2 *does*, effectively making an audit less useful and the value proposition worse.
That is why audits are generally only done on LTS or at least stable versions of software. It's just so much more cost-effective and audits are _expensive_.

As for closing this issue, that may be a good idea so long as it can be re-opened in the future. It's really up to you though :)
Thanks for looking into this, and I look forward to helping you with an audit when Reticulum has matured a bit more :heart:

**Comment by chron0** [comment]
Just adding my 2c here: 

I can certainly see that it makes sense to wait with an audit until things are more finalized, but it doesnt really hurt to prepare and plan an audit to be kicked off not long after. It seems to me that there is a lot of potential and also worldwide interest in reticulum, or more generally, decentralized p2p/mesh network solutions that also involve LoRa with COTS hw.

Therefore I expect more adoption and growth and this will need some trust, when we link up all the things and rely on an encrypted layer premise from a community perspective. If the audit finds potential snags or weaknesses in crypto implemenation (not even assumed intentional), it would be a much better outcome to have this a early as possible before people really start to rely on this. 
Secondly, maybe less likely but still not impossible, is the question if any findings would require some fundamental rethinks and redesigns on the crypto/protocol layer that would have further ripple effects.

And lastly, from experience with other mesh attempts over the last 2 decades: We really need as much resilience against interference as possible, from the RF or Network side. We will most definitely have to deal with bad actors trying to find ways to poison our networks, spam them, DoS them, on all available layers and vectors. Like in my personal case, i love the idea of rnsh but do i really trust it to run it or will i inadvertently open a vector for someone else to just take over my shell? The times where people are grateful for what they can use for free and therefore appreciate and dont mess with it, are long gone. With enough growth and accordingly perceived threat from it, regulators and governments will try as well. (see cryptocurrencies history).

So TL;DR - great job so far, thx a million for sharing - Audit Goals: 1. Acid Test Crypto 2. Acid Test Bad Actor Resilience

**Reply by cannon-c** [reply]
"The times where people are grateful for what they can use for free and therefore appreciate and dont mess with it, are long gone."

I am not sure what you mean by this sentence. But I agree with the rest of what you said.

**Reply by cannon-c** [reply]
Adding to what you said, I agree with you. Reticulum is a very needed thing, and we live in an adversarial world. Reticulum will be needed, and it must be assured to be as resilient as possible before then.

**Comment by cannon-c** [comment]
Closing this discussion for now, to re-open later (reference this conversation if this one cannot be re-opened) when Reticulum is matured more to be ready for a security audit.


This discussion has made progress though in preparation for such, including discussion on scope of an audit, and an idea on fund raising in a trustless and conditional way.
